pkg/util/tls.go

```diff
@@ -117,21 +180,30 @@ func NewTLSConfigFromServerConfiguration(configuration *configuration.ServerConf
 	}
 	tlsConfig.ClientAuth = tls.RequestClientCert
 
-	// Require the use of server-side certificates.
-	cert, err := tls.X509KeyPair([]byte(configuration.ServerCertificate), []byte(configuration.ServerPrivateKey))
-	if err != nil {
-		return nil, StatusWrapWithCode(err, codes.InvalidArgument, "Invalid server certificate or private key")
+	if configuration.ServerKeyPair == nil {
+		return nil, StatusWrapWithCode(err, codes.InvalidArgument, "Missing server_key_pair configuration")
 	}
-	tlsConfig.Certificates = []tls.Certificate{cert}
 
-	// Expose Prometheus metrics on certificate expiration.
-	leaf, err := x509.ParseCertificate(cert.Certificate[0])
-	if err != nil {
-		return nil, err
-	}
-	for _, dnsName := range leaf.DNSNames {
-		tlsServerCertificateNotBeforeTimeSeconds.WithLabelValues(dnsName).Set(float64(leaf.NotBefore.UnixNano()) / 1e9)
-		tlsServerCertificateNotAfterTimeSeconds.WithLabelValues(dnsName).Set(float64(leaf.NotAfter.UnixNano()) / 1e9)
+	// Require the use of server-side certificates.
+	switch keyPair := configuration.ServerKeyPair.KeyPair.(type) {
+	case *pb.X509KeyPair_Inline_:
+		var cert tls.Certificate
+		if cert, err = tls.X509KeyPair([]byte(keyPair.Inline.Certificate), []byte(keyPair.Inline.PrivateKey)); err != nil {
+			return nil, StatusWrapWithCode(err, codes.InvalidArgument, "Invalid server certificate or private key")
+		}
+		updateServerTLSCertificateExpiry(&cert)
```

Comment:
What is the difference between the client & server code? Is it just the Prometheus metrics that need to be used? What are your thoughts on  the existing metrics to `buildbarn_tls_certificate_not_before_time_seconds` (i.e., removing the `client` midfix)? We could just let that be the value of some kind of label that's attached to them.

Once that's done, it should be pretty trivial to move this code into a common implementation that both the client & server code can use.

Reply:
I'm ok to rename the metric, I was approaching it this way to avoid a more annoying breaking change by renaming the metric. If thats not a concern, then I will update to use a metric with label `certificate_usage=server|client`

Reply:
Sounds perfect!